### ISSUE TITLE
[Bug] Nested `typing.NewType` type hints cause @beartype to punch itself in the head.

### ISSUE THREAD

**Issue by rlkelly** [beartype/beartype#275]
If I run this:

```
from typing import NewType

from beartype import beartype


First = NewType("First", str)
Second = NewType("Second", First)

@beartype
def foo(x: Second):
    print(x)

```

I get the error: `beartype.roar.BeartypeDecorHintPepUnsupportedException: function __main__.foo() parameter "x" type hint __main__.First unsupported but erroneously detected as supported as beartype sign "HintSignNewType".`

I'm not sure what the root cause of this is, nor why this causes a `BeartypeDecorHintPepUnsupportedException`.

**Comment by leycec** [comment]
**Wowza!** That's a really fascinating edge case. I *never* thought to try nesting `NewType` declarations. Do static type-checkers like mypy and `pyright` support that, too? I bet they do. It's only @beartype that's out here failing its userbase. The shame I feel is real. :weary: 

> I'm not sure what the root cause of this is, nor why this causes a `BeartypeDecorHintPepUnsupportedException`.

The `BeartypeDecorHintPepUnsupportedException` exception you're seeing is, indeed, exceedingly rare. It happens when @beartype confuses itself by punching itself in the head. A high-level layer of the @beartype codebase says, "This thing is supported! Let's type-check this." That layer then passes that type hint (in this case, your nested `Second` type hint) onto another low-level layer of the @beartype codebase, which then says, "This thing isn't supported at all! What the heck, high-level layer? Let's vomit in our mouths and raise an exception."

Good times, @beartype codebase. Thanks for the excellent issue, @rlkelly. We'll get this obscure insanity patched up in time for our upcoming `beartype 0.16.0` release.

**Comment by rlkelly** [comment]
I'm excited to see your github responses are as animated as your comments in the codebase!  I spent a few hours on this until i realized it was time to make a MRE and confirm it wasn't my lack of understanding but an actual bug.  Anyways, I appreciate your responsiveness, I will use classes to represent the type hierarchy until this is fixed.

**Comment by rlkelly** [comment]
PS. mypy had no issue with this, but there's no reason for shame.  For some context, the reason I ran into this is building string types of increasing specificity.  As a toy example:

```python
Username = NewType("Username", str)
CapitalUsername = NewType("CapitalUsername", Username)

user: Username = Username("leycec")
capital_username: CapitalUsername = CapitalUsername(user.upper())
```

as a footnote, I'd love to find a way to enforce some rules associated with these types.  I realize this would break the constant time evaluation promise of beartype, but being able to ensure that `Username` doesn't have any spaces or `CapitalUsername` is a valid username that is all caps would be really powerful for the safety of a codebase.  If you have any recommendation on enforcement of `NewType`s, especially in relation to strings, I'd love to hear your feedback.

**Comment by leycec** [comment]
**Hah-ha!** @beartype actually did something that works. Shockingly, what you want [already exists under the `beartype.vale` (i.e., @beartype validator) API](https://beartype.readthedocs.io/en/latest/api_vale/#trendy-string-matching). Let's type this:

```python
from beartype import beartype
from beartype.vale import Is
from beartype.typing import Annotated, NewType

# Custom string "subclass" matching only non-empty strings.
nonempty_str = NewType(
    'nonempty_str', Annotated[str, Is[lambda text: bool(text)]])

@beartype
def foo(x: nonempty_str) -> None:
    print(x)

foo(nonempty_str('This is good.'))
foo(nonempty_str(''))
```

...which prints:

```python
This is good.
Traceback (most recent call last):
  File "/home/leycec/tmp/mopy.py", line 15, in <module>
    foo(nonempty_str(''))
  File "<@beartype(__main__.foo) at 0x7f7a0e8fa700>", line 28, in foo
beartype.roar.BeartypeCallHintParamViolation: Function __main__.foo() parameter x='' violates type hint __main__.nonempty_str,
as str '' violates validator Is[lambda text: bool(text)]:
    False == Is[lambda text: bool(text)].
```

Admittedly, that exception message could be somewhat more readable. But let's pretend this already does everything you've ever desired out of a validation framework. Flex those berry stronk biceps, @beartype! :muscle: :bear:

**Comment by leycec** [comment]
**Resolved by d369f5454fd39f.** Thanks so much for *bearing* with us, @rlkelly. <sup>hyuk hyuk</sup>

Let us know if you hit any other stumbling blocks! We'll promptly smash them with our Keyboard of Arcane Canadian Power + 1.
